File: 0001-0500/0359. Logger Rate Limiter.py
# Easy
# Map, Design
# Time:O(1), Space:O(n) 
# https://leetcode.cn/problems/logger-rate-limiter/

class Logger:
    def __init__(self):
        self.log_dict = {} # msg -> last_ts
    
    def shouldPrintMessage(self, timestamp: int, message: str) -> bool:
        # if not in dict -> add and print
        if message not in self.log_dict:
            self.log_dict[message] = timestamp
            return True
        
        if timestamp - self.log_dict[message] >= 10:
            self.log_dict[message] = timestamp
            return True
        
        return False

File: 0001-0500/0359. Logger Rate Limiter.go
package main
// Easy
// Time:O(1), Space:O(n)
// https://leetcode.cn/problems/logger-rate-limiter/

// Need to store message and timestamp, and need O(1) lookup and update -> HashMap

type Logger struct {
    lastPrint map[string]int
}


func Constructor() Logger {
    return Logger{
        lastPrint: make(map[string]int),
    }
}


func (this *Logger) ShouldPrintMessage(timestamp int, message string) bool {
    if lastTime, exists := this.lastPrint[message]; exists {
        if timestamp - lastTime < 10 {
            return false
        }
    }
    this.lastPrint[message] = timestamp
    return true
}
